### PR TITLE
Do not rebuild packages that have been already generated if the -c option is passed

### PR DESCRIPTION
**Pull request by esteve**
See https://github.com/jspricke/ros-deb-builder-action/issues/5#issuecomment-1415224100

### CODE REVIEW

**Comment by jspricke** [comment]
Thanks! I reworked your commit to use `getopts` due to being POSIX and a bit easier to read. I also changed it to use short opts (-c), I don't think we need the long versions for now. Also changed the skip logic to not need to indent all code and fix some corner cases with it.
Also I added a commit to simplify local usage, which I think fits here. Can you test it?

**Comment by esteve** [comment]
@jspricke thanks for suggestions, but I reverted the latest changes as they were out of scope for this PR. I've incorporated your changes for `getopts`, replaced `stat` with another conditional to keep it POSIX and added your instructions to README.md

I've fixed some spelling mistakes and rephrased the instructions to make them clearer, but I can't figure out what you meant with this line:

https://github.com/jspricke/ros-deb-builder-action/pull/10/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R57

Could you elaborate? I can rephrase it to sound more natural / clearer.

One drawback of using `getopts` is that this line won't seem to work, because there's no parsing of `--` to separate external arguments from internal ones (e.g. if the user wants to pass anything down to `sbuild`)

https://github.com/esteve/ros-deb-builder-action/blob/continue-packages/build.sh#L103

Lastly, it's generally frowned upon to force-push onto another user's fork as it'll invalidate their local copy. Please instead open a new PR to discuss the changes or add the changes on top of the existing commits, I've added your changes as a separate commit after mine.

**Comment by jspricke** [comment]
Hi Esteve,

>@jspricke thanks for suggestions, but I reverted the latest changes as they were out of scope for this PR.

Sure, but note that those where my proposals how to get your changes in ;).

>I've incorporated your changes for `getopts`, replaced `stat` with another conditional to keep it POSIX and added your instructions to README.md

Did you test this? What happens if the glob matches no file or more than 
one?

>I've fixed some spelling mistakes and rephrased the instructions to make them clearer, but I can't figure out what you meant with this line:
>
>https://github.com/jspricke/ros-deb-builder-action/pull/10/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R57
>
>Could you elaborate? I can rephrase it to sound more natural / clearer.

rosdep update is modified here:

https://github.com/jspricke/ros-deb-builder-action/blob/continue-packages/build.sh#L70

and reverted here:

https://github.com/jspricke/ros-deb-builder-action/blob/continue-packages/build.sh#L116

So in between rosdep can return different results.

>One drawback of using `getopts` is that this line won't seem to work, because there's no parsing of `--` to separate external arguments from internal ones (e.g. if the user wants to pass anything down to `sbuild`)
>
>https://github.com/esteve/ros-deb-builder-action/blob/continue-packages/build.sh#L103

Can you give an example? It works for me.

>Lastly, it's generally frowned upon to force-push onto another user's fork as it'll invalidate their local copy. Please instead open a new PR to discuss the changes or add the changes on top of the existing commits, I've added your changes as a separate commit after mine.

There is a button in Github to disable that if you can't handle it.

Cheers Jochen

**Comment by esteve** [comment]
> Sure, but note that those where my proposals how to get your changes in ;).

I think we have a very different definition of what a suggestion is, I know you can definitely know how to use reviews, and that'd be the place for suggestions.

> Did you test this? What happens if the glob matches no file or more than
one?

I have, it works when no files match, haven't tested when there's more than one file, that'd be interesting to test, but probably won't work as an if clause, I agree.

Anyway, back to `getopt`/ `getopts`, I thought the point of using `getopts` instead of `getopt` was for POSIX compliance, if you're ok with `stat`, then `getopt` should be fine too.

> rosdep update is modified here: https://github.com/jspricke/ros-deb-builder-action/blob/continue-packages/build.sh#L70 and reverted here: https://github.com/jspricke/ros-deb-builder-action/blob/continue-packages/build.sh#L116 So in between rosdep can return different results.

Ah, I see now what you meant. I can rephrase that sentence to make it more understandable, thanks.

> Can you give an example? It works for me.

You're right, I missed the `shift` in your changes. See? Now we're discussing contributions.

> There is a button in Github to disable that if you can't handle it.

Not sure what you mean by "handle it", there's no need to be antagonistic. I always leave that option on because I'm open to maintainers pushing commits onto my fork, but forcing commits so that my local copy diverges and I can't pull cleanly from the remote is definitely not cool, and I'm sure you'd agree.

**Comment by esteve** [comment]
@jspricke I've updated the note about `rosdep` in 8ea7ac0, is that what you meant?

**Comment by jspricke** [comment]
>I think we have a very different definition of what a suggestion is, I know you can definitely know how to use reviews, and that'd be the place for suggestions.

Sure but it would have taken more time to add them as suggestions and I 
don't think it is worth the time.

>Anyway, back to `getopt`/ `getopts`, I thought the point of using `getopts` instead of `getopt` was for POSIX compliance, if you're ok with `stat`, then `getopt` should be fine too.

If you have a proposal to implement in POSIX without being unreadable, 
I'm all ears.
Do you have any reason to prefer getopt?

>@jspricke I've updated the note about `rosdep` in 8ea7ac0, is that what you meant?

I'm not sure that makes it easier to understand. What do you mean by 
"inconsistent" in that context? The cache is definitely changed, so it 
is not conditional on an external(?) change.

**Comment by esteve** [comment]
> Sure but it would have taken more time to add them as suggestions and I don't think it is worth the time.

To make my position clear: commits on top of mine yes, force-push that rewrite existing commits that make brances diverge nope. I've reenabled edits to this PR, so feel free to add more commits on top.

Anyway, back to the changes themselves.

> If you have a proposal to implement in POSIX without being unreadable,
> I'm all ears.
> Do you have any reason to prefer getopt?

I find `getopt` more readable than `getopts`, I just think the POSIX compliance argument inconsistent. `stat` is not POSIX, and is ok, but `getopt` is not. I'd prefer that you make it clear that it's not about POSIX-ness, but down to personal preferences (which is totally fine, nothing to argue about that).

> I'm not sure that makes it easier to understand. What do you mean by "inconsistent" in that context? The cache is definitely changed, so it is not conditional on an external(?) change.

In that case, I didn't understand the intent of the original comment, so it proves that it definitely needed clarification. What you meant originally is that the `rosdep` cache first adds the local repo as a source and then discards it to go back to the main ROS index?

File: build.sh [hunk 1]
```diff
@@ -85,7 +85,7 @@ for PKG_PATH in $(catkin_topological_order --only-folders); do
   (
   cd "$PKG_PATH"
   test -f debian/changelog && GENERATED_DEBIAN_PACKAGE=$(head -n1 debian/changelog | cut -d' ' -f1)
-  if [ -n "$CONTINUE_PACKAGE_GENERATION" ] && [ -n "$GENERATED_DEBIAN_PACKAGE" ] && stat -t "${APT_REPO}/${GENERATED_DEBIAN_PACKAGE}_"*.deb ]; then
+  if [ -n "$CONTINUE_PACKAGE_GENERATION" ] && [ -n "$GENERATED_DEBIAN_PACKAGE" ] && ls "${APT_REPO}/${GENERATED_DEBIAN_PACKAGE}_"*.deb >/dev/null 2>&1; then
```
Comment (esteve) [inline build.sh:88]: @jspricke regarding POSIX-ness, I've replaced `stat` with `ls`, this should be roughly equivalent, let me know if you can think of any unintentional side effects

File: README.md [hunk 2]
```diff
@@ -53,6 +53,6 @@ Run `prepare.sh` once to set up the system and/or adopt to your needs, it is adv
 It will create a `~/.cache/sbuild` with sufficient rights for the sbuild process, a `~/.sbuildrc` and a `./src` with the checked out repos.
 Run `build.sh` in a ROS workspace (or the current directory with `./src`) to generate the packages into the `apt_repo` folder.
 You can run `build.sh -c` to skip already built packages.
-The `rosdep` cache is updated twice within the script, so results might be inconsistent if there's a change in the ROS index.
+The `rosdep` cache is updated twice within the script to include the packages in the local repo, but reverted to the original cache afterwards.
```
[the remark this answers is not in the document]
Comment (jspricke) [reply README.md:56]: works for me, thx.

File: build.sh (hunk 1 above)
Comment (jspricke) [reply build.sh:88]: Ha, I guess I was too tired last night, thx.

File: README.md [hunk 3]
```diff
@@ -45,3 +45,14 @@ with:
   DEB_DISTRO: jammy
   GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
 ```
+
+## Running manually
```
Comment (jspricke) [inline README.md:49]: this paragraph seems out of scope as you removed the commit that it belong to. Can you remove it?

File: build.sh [hunk 4]
```diff
@@ -3,13 +3,38 @@
...
+      ;;
+    *)
+      echo "Error parsing arguments"
+      exit 1
+      ;;
+  esac
+done
+shift "$((OPTIND - 1))"
+
+# make output directory
```
Comment (jspricke) [inline build.sh:24]: dito this part.

File: build.sh (hunk 1 above)
Comment (esteve) [reply build.sh:88]: No problem, glad we found a way :slightly_smiling_face:

File: build.sh [hunk 5]
```diff
@@ -34,18 +59,15 @@ case $ROS_DISTRO in
...
 esac
 
-# make output directory
-mkdir /home/runner/apt_repo
-
 echo "Add unreleased packages to rosdep"
 
 for PKG in $(catkin_topological_order --only-names); do
-  printf "%s:\n  %s:\n  - %s\n" "$PKG" "$DISTRIBUTION" "ros-$ROS_DEB$(printf '%s' "$PKG" | tr '_' '-')" >> /home/runner/apt_repo/local.yaml
+  printf "%s:\n  %s:\n  - %s\n" "$PKG" "$DISTRIBUTION" "ros-$ROS_DEB$(printf '%s' "$PKG" | tr '_' '-')" >> ${APT_REPO}/local.yaml
```
Comment (jspricke) [inline build.sh:65]: dito this.

btw. please don't copy parts of a commit from a different author out of context without proper attribution (use `git commit --author`, maybe)

File: README.md [hunk 6]
```diff
@@ -45,3 +45,14 @@ with:
...
+
+You can run this action locally on Ubuntu 22.04 ("jammy") or Debian 12 ("bookworm") and newer.
+Run `prepare.sh` once to set up the system and/or adopt to your needs, it is advised to run it as a separate user to avoid polluting your home directory.
```
Comment (esteve) [inline README.md:52]: @jspricke did you mean adopt here or adapt?

File: build.sh [hunk 7]
```diff
@@ -62,6 +84,13 @@ for PKG_PATH in $(catkin_topological_order --only-folders); do
   test -f "$PKG_PATH/COLCON_IGNORE" && echo "Skipped" && continue
   (
   cd "$PKG_PATH"
+  test -f debian/changelog && GENERATED_DEBIAN_PACKAGE=$(head -n1 debian/changelog | cut -d' ' -f1)
```
Comment (jspricke) [inline build.sh:87]: `debian/changelog` does not exist at this point. Feel free to revert to my proposed change. For reference: https://github.com/jspricke/ros-deb-builder-action/commit/4c058db5d37e26fadff10e66f58e48257c0aa867#diff-4d2a8eefdf2a9783512a35da4dc7676a66404b6f3826a8af9aad038722da6823R83

File: build.sh [hunk 8]
```diff
@@ -62,6 +84,13 @@ for PKG_PATH in $(catkin_topological_order --only-folders); do
...
+  if [ -n "$CONTINUE_PACKAGE_GENERATION" ] && [ -n "$GENERATED_DEBIAN_PACKAGE" ] && ls "${APT_REPO}/${GENERATED_DEBIAN_PACKAGE}_"*.deb >/dev/null 2>&1; then
+    echo " Skipping already generated package: ${GENERATED_DEBIAN_PACKAGE}"
+    COUNT=$((COUNT+1))
```
Comment (jspricke) [inline build.sh:90]: this is not needed.

File: build.sh [hunk 9]
```diff
@@ -62,6 +84,13 @@ for PKG_PATH in $(catkin_topological_order --only-folders); do
...
+    echo "::endgroup::"
+    continue
```
Comment (jspricke) [inline build.sh:92]: continue does not work in a subshell.

File: build.sh [hunk 10]
```diff
@@ -74,11 +103,14 @@ for PKG_PATH in $(catkin_topological_order --only-folders); do
 
   # dpkg-source-opts: no need for upstream.tar.gz
   sbuild --chroot-mode=unshare --no-clean-source --no-run-lintian \
-    --dpkg-source-opts="-Zgzip -z1 --format=1.0 -sn" --build-dir=/home/runner/apt_repo \
-    --extra-package=/home/runner/apt_repo "$@"
+    --dpkg-source-opts="-Zgzip -z1 --format=1.0 -sn" --build-dir=${APT_REPO} \
```
Comment (jspricke) [inline build.sh:106]: revert this as well, please.

File: build.sh [hunk 11]
```diff
@@ -74,11 +103,14 @@ for PKG_PATH in $(catkin_topological_order --only-folders); do
...
+    --extra-package=${APT_REPO} "$@"
   )
   COUNT=$((COUNT+1))
   echo "::endgroup::"
 done
 
+# revert rosdep cache
```
Comment (jspricke) [inline build.sh:113]: dito this.

**Review by jspricke — CHANGES_REQUESTED** [review]

File: README.md (hunk 6 above)
Comment (jspricke) [reply README.md:52]: adapt, thx.

File: build.sh (hunk 5 above)
Comment (esteve) [reply build.sh:65]: Didn't mean to steal your attribution at all, sorry. `git commit --author` is fine and I can add you in the commit message if that's what do you prefer.

File: build.sh (hunk 7 above)
Comment (esteve) [reply build.sh:87]: It should if it's been previously generated, right? And if not, that means we'll have to generate the package, wouldn't it? I prefer using `head` instead of `catkin_topological_order`, as it should be quicker.

Comment (jspricke) [reply build.sh:87]: that's a bold assumption, I would not follow that.

File: build.sh (hunk 9 above)
Comment (esteve) [reply build.sh:92]: It seems to be working fine locally, though:

```shell
...
+ test -f src/universe/autoware.universe/system/default_ad_api_helpers/ad_api_visualizers/CATKIN_IGNORE
+ test -f src/universe/autoware.universe/system/default_ad_api_helpers/ad_api_visualizers/COLCON_IGNORE
+ cd src/universe/autoware.universe/system/default_ad_api_helpers/ad_api_visualizers
+ test -f debian/changelog
+ head -n1 debian/changelog
+ cut -d  -f1
+ GENERATED_DEBIAN_PACKAGE=ros-humble-ad-api-visualizers
+ [ -n 1 ]
+ [ -n ros-humble-ad-api-visualizers ]
+ ls /home/runner/ros-deb-builder-action/apt_repo/ros-humble-ad-api-visualizers_0-2023.02.09.15.40_amd64.deb
+ echo  Skipping already generated package: ros-humble-ad-api-visualizers
 Skipping already generated package: ros-humble-ad-api-visualizers
+ COUNT=2
+ echo ::endgroup::
::endgroup::
+ continue
+ COUNT=2
+ echo ::endgroup::
::endgroup::
+ echo ::group::Building 2/235: src/core/autoware_adapi_msgs/autoware_adapi_v1_msgs
::group::Building 2/235: src/core/autoware_adapi_msgs/autoware_adapi_v1_msgs
+ test -f src/core/autoware_adapi_msgs/autoware_adapi_v1_msgs/CATKIN_IGNORE
+ test -f src/core/autoware_adapi_msgs/autoware_adapi_v1_msgs/COLCON_IGNORE
+ cd src/core/autoware_adapi_msgs/autoware_adapi_v1_msgs
+ test -f debian/changelog
+ head -n1 debian/changelog
+ cut -d  -f1
+ GENERATED_DEBIAN_PACKAGE=ros-humble-autoware-adapi-v1-msgs
+ [ -n 1 ]
+ [ -n ros-humble-autoware-adapi-v1-msgs ]
+ ls /home/runner/ros-deb-builder-action/apt_repo/ros-humble-autoware-adapi-v1-msgs_0-2023.02.09.15.41_amd64.deb
+ echo  Skipping already generated package: ros-humble-autoware-adapi-v1-msgs
 Skipping already generated package: ros-humble-autoware-adapi-v1-msgs
+ COUNT=3
...
```

or you mean something else?

Comment (jspricke) [reply build.sh:92]: Afaik that depends on the shell implementation and `exit` should always be correct.

File: build.sh (hunk 10 above)
Comment (esteve) [reply build.sh:106]: Can you elaborate why it should be reverted? Or you mean attribution? I think it's correct, but I'll update the author to point to you.

File: build.sh (hunk 11 above)
Comment (esteve) [reply build.sh:113]: Same question as https://github.com/jspricke/ros-deb-builder-action/pull/10#discussion_r1101709187

PS: it's ditto, with two "t" ;-)

File: build.sh (hunk 7 above)
Comment (jspricke) [reply build.sh:87]: Well all of it. I assume that the user can do anything between two runs, including deleting `debian/changelog` and I assume their computer is fast enough that the different commands do not matter.

File: build.sh (hunk 11 above)
Comment (jspricke) [reply build.sh:113]: Interesting, it is dito or detto in German and French (according to Wikipedia) but ditto in Italian and English, thx :).

File: build.sh (hunk 7 above)
Comment (esteve) [reply build.sh:87]: That's an interesting case, I'm sure the user can do plenty of other things between two runs that may conflict with this script, but I'm not stubborn enough that this change bothers me that much, I'll update it. We could be arguing ad nauseam, which would be a waste of time.

File: README.md (hunk 3 above)
Comment (esteve) [reply README.md:49]: Not sure what you mean here. Remove the whole paragraph? The line preceding it?

File: README.md [hunk 12]
```diff
@@ -45,3 +45,14 @@ with:
...
+It will create a `~/.cache/sbuild` with sufficient rights for the sbuild process, a `~/.sbuildrc` and a `./src` with the checked out repos.
+Run `build.sh` in a ROS workspace (or the current directory with `./src`) to generate the packages into the `apt_repo` folder.
+You can run `build.sh -c` to skip already built packages.
+The `rosdep` cache is updated twice within the script to include the packages in the local repo, but reverted to the original cache afterwards.
+Run `repository.sh` to create an APT repository.
+You can directly use it on your local machine by adopting the path from the generated `README.md`.
```
Comment (esteve) [inline README.md:58]: Is it adopting or adapting here?

File: README.md (hunk 3 above)
Comment (jspricke) [reply README.md:49]: All current changes in the `README.md` as they describe the state after my commit and are unrelated to this PR.

File: README.md (hunk 12 above)
Comment (jspricke) [reply README.md:58]: adapt as well, thx.

File: build.sh (hunk 9 above)
Comment (esteve) [reply build.sh:92]: As you surely, `/bin/sh` on Debian points to `dash`, which runs as POSIX dash, so if this passes with `/bin/sh`, it should be safe to assume that `continue` is fine, wouldn't it? I might be mistaken here, though.

File: build.sh (hunk 10 above)
Comment (esteve) [reply build.sh:106]: You mean the change replacing `/home/runner/apt_repo` with `${APT_REPO}`, right?

Comment (jspricke) [reply build.sh:106]: yes.

File: build.sh (hunk 9 above)
Comment (jspricke) [reply build.sh:92]: What's wrong with using exit here?

File: build.sh (hunk 11 above)
Comment (esteve) [reply build.sh:113]: Ah, the original is Italian (because `dito` means finger), but I guess German wouldn't have that problem so I presume that's why `dito` is correct. Interesting there's a `detto` variant.

File: build.sh [hunk 13]
```diff
@@ -62,6 +80,11 @@ for PKG_PATH in $(catkin_topological_order --only-folders); do
   test -f "$PKG_PATH/COLCON_IGNORE" && echo "Skipped" && continue
   (
   cd "$PKG_PATH"
+  GENERATED_DEBIAN_PACKAGE=ros-${ROS_DEB}$(catkin_topological_order --only-names | tr '_' '-')
+  if [ -n "$CONTINUE_PACKAGE_GENERATION" ] && [ -n "$GENERATED_DEBIAN_PACKAGE" ] && ls "/home/runner/apt_repo/${GENERATED_DEBIAN_PACKAGE}_"*.deb >/dev/null 2>&1; then
```
Comment (jspricke) [inline build.sh:84]: the variable and the extra check seems overkill.

Comment (esteve) [reply build.sh]: Removed

File: build.sh [hunk 14]
```diff
@@ -62,6 +80,11 @@ for PKG_PATH in $(catkin_topological_order --only-folders); do
   test -f "$PKG_PATH/COLCON_IGNORE" && echo "Skipped" && continue
   (
   cd "$PKG_PATH"
+  GENERATED_DEBIAN_PACKAGE=ros-${ROS_DEB}$(catkin_topological_order --only-names | tr '_' '-')
```
Comment (jspricke) [inline build.sh:83]:
```suggestion
  GENERATED_DEBIAN_PACKAGE="ros-${ROS_DEB}$(catkin_topological_order --only-names | tr '_' '-')"
```

**Comment by jspricke** [comment]
LGTM, can you please squash your commits.

**Comment by esteve** [comment]
@jspricke if I squash them your attribution is gonna get lost, you ok with that?

**Comment by jspricke** [comment]
sure.

**Comment by esteve** [comment]
@jspricke done, I kept the Co-authored line so that you get credited as well.

**Comment by jspricke** [comment]
Can you please shorten the commit message a bit?

**Comment by esteve** [comment]
@jspricke done

**Comment by esteve** [comment]
@jspricke it's been a pleasure, thanks for taking the time to review this PR
